src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -3379,87 +3352,84 @@ void EmitLazy(RegexNode node)
-                    // else will overwrite them between now and when backtracking occurs.  If, however, we are inside
-                    // of another loop, then any number of iterations might have such state that needs to be stored,
-                    // and thus it needs to be pushed on to the backtracking stack.
-                    if (rm.Analysis.IsInLoop(node))
-                    {
-                        EmitStackPush(iterationMayBeEmpty ?
-                            new[] { iterationCount, startingPos!, sawEmpty! } :
-                            new[] { iterationCount });
-                    }
 
-                    string skipBacktrack = ReserveName("LazyLoopSkipBacktrack");
-                    Goto(skipBacktrack);
-                    writer.WriteLine();
+                // We need to ensure that some state (e.g. iteration count) is persisted if we're backtracked to.
+                // We also need to push the current position, so that subsequent iterations pick up at the right
+                // point (and subsequent expressions are almost certain to have changed the current pos). However,
+                // if we're not inside of a loop, the other local's used for this construct are sufficient, as nothing
+                // else will overwrite them between now and when backtracking occurs.  If, however, we are inside
+                // of another loop, then any number of iterations might have such state that needs to be stored,
+                // and thus it needs to be pushed on to the backtracking stack.
+                EmitStackPush(
+                    !rm.Analysis.IsInLoop(node) ? new[] { "pos" } :
+                    iterationMayBeEmpty ? new[] { "pos", iterationCount, startingPos!, sawEmpty! } :
+                    new[] { "pos", iterationCount });
 
-                    // Emit a backtracking section that checks the timeout, restores the loop's state, and jumps to
-                    // the appropriate label.
-                    string backtrack = ReserveName($"LazyLoopBacktrack");
-                    MarkLabel(backtrack, emitSemicolon: false);
+                string skipBacktrack = ReserveName("LazyLoopSkipBacktrack");
+                Goto(skipBacktrack);
+                writer.WriteLine();
 
-                    // We're backtracking.  Check the timeout.
-                    EmitTimeoutCheckIfNeeded(writer, rm);
+                // Emit a backtracking section that checks the timeout, restores the loop's state, and jumps to
+                // the appropriate label.
+                string backtrack = ReserveName($"LazyLoopBacktrack");
+                MarkLabel(backtrack, emitSemicolon: false);
 
-                    if (rm.Analysis.IsInLoop(node))
+                // We're backtracking.  Check the timeout.
+                EmitTimeoutCheckIfNeeded(writer, rm);
+
+                EmitStackPop(
+                    !rm.Analysis.IsInLoop(node) ? new[] { "pos" } :
```

Comment:
NIT: looks like we are checking the IsInLoop for this node twice. should we just instead cache it in a local instead?

Reply:
Yeah, might as well cache it to a local.